### PR TITLE
Optimize encoding and decoding of `Vec<T>` when `T` is trivial

### PR DESCRIPTION
**Pull request by ironcev**
## Description

This PR optimizes the encoding and decoding of `std::vec::Vec::<T>` when `T` is a trivially encodable and decodable type. In that case, we don't have encode each element individually, but just encode the vector's length and mem-copy its content to the `Buffer`. 

## Checklist

- [ ] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
   - [ ] If my change requires substantial documentation changes, I have [requested support from the DevRel team](https://github.com/FuelLabs/devrel-requests/issues/new/choose)
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Comment by ironcev** [comment]
:+1:

File: test/src/e2e_vm_tests/test_programs/should_pass/stdlib/vec_encoding_decoding/src/main.sw [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+fn vec_non_trivial() {
+    let vec = create_vec_trivial(10);
+    let encoded_decoded: Vec<u64> = abi_decode(encode(vec));
+    assert_eq(encoded_decoded, vec);
+
+    let encoded_as_alias = encode_allow_alias(&encoded_decoded);
+    let encoded_decoded: Vec<u64> = abi_decode(encoded_as_alias);
+    assert_eq(encoded_decoded, vec);
+
+    log(encoded_decoded);
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply test/src/e2e_vm_tests/test_programs/should_pass/stdlib/vec_encoding_decoding/src/main.sw:63]: Done.

File: test/src/e2e_vm_tests/test_programs/should_pass/stdlib/vec_encoding_decoding/test.run.toml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+category = "run"
+
+script_data_new_encoding = "0000000000000003000000000000007c000000000000007c000000000000007c 00000000000000030000007c0000007c0000007c"
+expected_result_new_encoding = { action = "return_data", value = "0000000000000006000000000000007c000000000000007c000000000000007c000000000000007c000000000000007c000000000000007c 00000000000000060000007c0000007c0000007c0000007c0000007c0000007c" }
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply test/src/e2e_vm_tests/test_programs/should_pass/stdlib/vec_encoding_decoding/test.run.toml:4]: Not relevant anymore.

**Comment by xunilrj** [comment]
Before this optimisation

```
test in_vec_trivial ... ok (???, 4860 gas)
```

After

```
 test in_vec_trivial ... ok (???, 2703 gas)
```

File: sway-lib-std/src/vec.sw [hunk 3]
```diff
@@ -933,3 +944,71 @@ where
         l.finish();
     }
 }
+
+struct NonTrivial {
```
Comment (ironcev) [inline sway-lib-std/src/vec.sw]: Because we currently don't support neither inline modules nor something like `#[cfg(test = true)` having test types and impls in modules bloats the module and is actually confusing 😞 I was confused at first when encountering `NonTrivial` I thought it is some actual type and not a test dummy.

That's why I would propose to move these tests to in-language tests: `test/src/in_language_tests/test_programs/vec_inline_tests/src/main.sw`

In general, we have a very very few tests in `std` modules and the actual `std` tests are in in-language tests. IMO, we should move them all to in-language-tests, but that's a topic for another day.

File: sway-lib-std/src/vec.sw [hunk 4]
```diff
@@ -933,3 +944,71 @@ where
...
+}
+
+impl PartialEq for NonTrivial {
+    fn eq(self, other: Self) -> bool {
+        self.v == other.v
+    }
+}
+
+#[test]
+fn vec_encode_trivial() {
```
Comment (ironcev) [inline sway-lib-std/src/vec.sw]: Nitpick. The body of `vec_encode_trivial` and `vec_encode_non_trivial` is essentially the same. We can have one generic `test_vec_encode<T>(v1: T, v2: T)` here and call it in both tests.

File: sway-lib-std/src/vec.sw [hunk 5]
```diff
@@ -933,3 +944,71 @@ where
...
+    }
+}
+
+impl PartialEq for NonTrivial {
+    fn eq(self, other: Self) -> bool {
+        self.v == other.v
+    }
+}
+
+#[test]
```
Comment (ironcev) [inline sway-lib-std/src/vec.sw]: A corner case that IMO should be covered in tests is having a **zero-sized** `T` once trivial, once non-trivial. Maybe just a reason more than for having a once defined `test_vec_encode<T>(v1: T, v2: T)` 🤔

File: sway-core/src/semantic_analysis/ast_node/expression/intrinsic_function.rs [hunk 6]
```diff
@@ -738,6 +738,28 @@ fn type_check_encode_append(
             | TypeInfo::StringArray(_)
             | TypeInfo::StringSlice
             | TypeInfo::RawUntypedSlice => {}
+            TypeInfo::Tuple(items) => {
+                if items.len() != 2 {
+                    return Err(
+                        handler.emit_err(CompileError::EncodingUnsupportedType { span: item_span })
```
Comment (ironcev) [inline sway-core/src/semantic_analysis/ast_node/expression/intrinsic_function.rs:744]: We need to update the error message in `error.rs` as well.

Proposal:
Only bool, u8, u16, u32, u64, u256, b256, string arrays, string slices, and (raw_ptr, u64) can be used here.

Comment (xunilrj) [reply sway-core/src/semantic_analysis/ast_node/expression/intrinsic_function.rs:744]: Done.

File: sway-lib-std/src/vec.sw (hunk 5 above)
Comment (xunilrj) [reply sway-lib-std/src/vec.sw]: Done at https://github.com/FuelLabs/sway/pull/7620

File: sway-lib-std/src/raw_slice.sw [hunk 7]
```diff
@@ -164,7 +164,12 @@ impl raw_slice {
     /// }
     /// ```
     pub fn len<T>(self) -> u64 {
-        into_parts(self).1 / __size_of::<T>()
+        let len = __size_of::<T>();
+        if len != 0 {
+            into_parts(self).1 / len
+        } else {
+            __revert(999);
```
Comment (ironcev) [inline sway-lib-std/src/raw_slice.sw]: Everywhere in the `std` we have `[_]revert(0)`, except for the very few `.._SIGNAL` constants.

This is IMO one of the most problematic decisions in the current error handling, complete lack of any additional information on most of the reverts 😞 

From that perspective, I am more for `999` than `0`, but breaking the pattern only here and not having `999` documented anywhere nor following any convention is also not bringing much. Also, we will be replacing all `revert(0)`s with `panic` (hopefully) soon.

That's why my proposal here is to stick to the pattern we have everywhere else in the `std`.

I would leave the `if` still, just to know we need to handle it later with `panic`.
```suggestion
            __revert(0);
```

File: sway-lib-std/src/raw_slice.sw
Comment (ironcev) [inline sway-lib-std/src/raw_slice.sw:155]: We need to document the revert condition.
```suggestion
    ///
    /// # Reverts
    ///
    /// * When `T` is a zero-sized type.
    ///
```

File: sway-lib-std/src/raw_slice.sw (hunk 7 above)
Comment (xunilrj) [reply sway-lib-std/src/raw_slice.sw]: Done.

File: sway-lib-std/src/raw_slice.sw
Comment (xunilrj) [reply sway-lib-std/src/raw_slice.sw:155]: Done.

**Comment by xunilrj** [comment]
@cursor BugBot run

**Review by xunilrj — APPROVED** [review]
